src/rpc.rs

```diff
@@ -0,0 +1,364 @@
+//! The `rpc` module implements the Solana rpc interface.
+
+use bank::Bank;
+use bs58;
+use jsonrpc_core::*;
+use jsonrpc_http_server::*;
+use request::{Request as JsonRpcRequest, Response};
+use service::Service;
+use signature::{Pubkey, Signature};
+use std::mem;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::thread::{self, Builder, JoinHandle};
+
+pub const RPC_PORT: u16 = 8899;
+
+pub struct JsonRpcService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl JsonRpcService {
+    pub fn new(bank: Arc<Bank>, rpc_addr: SocketAddr) -> Self {
+        let request_processor = JsonRpcRequestProcessor::new(bank);
+        let thread_hdl = Builder::new()
+            .name("solana-jsonrpc".to_string())
+            .spawn(move || {
+                let mut io = MetaIoHandler::default();
+                let rpc = RpcSolImpl;
+                io.extend_with(rpc.to_delegate());
+
+                let server = ServerBuilder::new(io)
+                    .meta_extractor(move |_req: &hyper::Request| Meta {
+                        request_processor: Some(request_processor.clone()),
+                    })
+                    .threads(4)
+                    .cors(DomainsValidation::AllowOnly(vec![
+                        AccessControlAllowOrigin::Any,
+                    ]))
+                    .start_http(&rpc_addr)
+                    .unwrap();
+                server.wait();
+                ()
+            })
+            .unwrap();
+        JsonRpcService { thread_hdl }
+    }
+}
+
+impl Service for JsonRpcService {
+    fn thread_hdls(self) -> Vec<JoinHandle<()>> {
+        vec![self.thread_hdl]
+    }
+
+    fn join(self) -> thread::Result<()> {
+        self.thread_hdl.join()
+    }
+}
+
+#[derive(Clone)]
+pub struct Meta {
+    pub request_processor: Option<JsonRpcRequestProcessor>,
+}
+impl Metadata for Meta {}
+impl Default for Meta {
```

Comment:
Does some library require we implement `Default`? If not, it'd be better to get rid of that `Option<>` and pass the request processor into a `Meta` constructor. Then you'd be able to get rid of all those `unwrap()` calls on `request_processor`.

Reply:
The version of jsonrpc on crates.io requires `Default` for Metadata, but that has been fixed in more recent commits. I have changed the dependency to pull from github for now.